File: source/python/buffer_builder.py
# Growing buffer that doubles capacity when full

capacity = 4
buf = bytearray(capacity)
size = 0

def buf_grow():
    global capacity, buf
    new_cap = capacity * 2
    new_buf = bytearray(new_cap)
    new_buf[:size] = buf[:size]
    buf = new_buf
    capacity = new_cap

def write_byte(b):
    global size
    if size >= capacity:
        buf_grow()
    buf[size] = b
    size += 1

def read_back():
    return buf[:size].decode("utf-8")

message = "Hello, World!"
for b in message.encode("utf-8"):
    write_byte(b)

print(f"wrote {size} bytes")
print(f"capacity: {capacity}")
reconstructed = read_back()
print(f"reconstructed: {reconstructed}")

File: source/go/buffer_builder.go
package main

import "fmt"

// Growing buffer that doubles when full
var capacity = 4
var size = 0
var buf []byte

func bufGrow() {
	newCap := capacity * 2
	newBuf := make([]byte, newCap)
	copy(newBuf, buf)
	buf = newBuf
	capacity = newCap
}

func writeByte(b byte) {
	if size >= capacity {
		bufGrow()
	}
	buf[size] = b
	size++
}

func readBack() string {
	return string(buf[:size])
}

func main() {
	buf = make([]byte, capacity)

	message := "Hello, World!"
	for i := 0; i < len(message); i++ {
		writeByte(message[i])
	}

	fmt.Printf("wrote %d bytes\n", size)
	fmt.Printf("capacity: %d\n", capacity)

	reconstructed := readBack()
	fmt.Printf("reconstructed: %s\n", reconstructed)
}
